linfa-clustering/src/optics/algorithm.rs

```diff
@@ -0,0 +1,247 @@
+    r_distance: Option<FloatOrd<f64>>,
+}
+
+impl<'a, F: Float> Neighbor<'a, F> {
+    /// Create a new neighbor
+    fn new(index: usize, observation: ArrayView<'a, F, Ix1>) -> Self {
+        Self {
+            index,
+            observation,
+            processed: false,
+            c_distance: None,
+            r_distance: None,
+        }
+    }
+
+    /// Set the core distance given the minimum points in a cluster and the points neighbors
+    fn set_core_distance(&mut self, min_pts: usize, neighbors: &[Neighbor<'a, F>]) {
+        self.c_distance = neighbors
+            .iter()
+            .nth(min_pts - 1)
+            .map(|x| FloatOrd(self.observation.l2_dist(&x.observation).unwrap()))
+            .clone();
+    }
+
+    /// Convert the neighbor to a sample for the user
+    fn sample(&self) -> Sample<'a, F> {
+        Sample {
+            index: self.index,
+            observation: self.observation.clone(),
+            reachability_distance: self.r_distance.map(|x| x.0),
+            core_distance: self.c_distance.map(|x| x.0),
+        }
+    }
+}
+
+impl<'a, F: Float> Eq for Neighbor<'a, F> {}
+
+impl<'a, F: Float> PartialEq for Neighbor<'a, F> {
+    fn eq(&self, other: &Self) -> bool {
+        self.r_distance == other.r_distance
+    }
+
+    fn ne(&self, other: &Self) -> bool {
+        !self.eq(other)
+    }
+}
+
+impl<'a, F: Float> PartialOrd for Neighbor<'a, F> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        self.r_distance.partial_cmp(&other.r_distance)
+    }
+}
+
+impl<'a, F: Float> Ord for Neighbor<'a, F> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.r_distance.cmp(&other.r_distance)
+    }
+}
+
+impl Optics {
+    pub fn params(min_points: usize) -> OpticsHyperParamsBuilder {
+        OpticsHyperParams::new(min_points)
+    }
+}
+
+impl<'a, F: Float, D: Data<Elem = F>> Transformer<&'a ArrayBase<D, Ix2>, OpticsResult<'a, F>>
+    for OpticsHyperParams
+{
+    fn transform(&self, observations: &'a ArrayBase<D, Ix2>) -> OpticsResult<'a, F> {
+        let mut result = OpticsResult { orderings: vec![] };
+
+        let mut points = observations
+            .axis_iter(Axis(0))
+            .enumerate()
+            .map(|(i, x)| Neighbor::new(i, x))
+            .collect::<Vec<_>>();
+
+        let mut processed = HashSet::new();
+        while processed.len() != points.len() {
+            let obs = points
+                .iter_mut()
+                .enumerate()
+                .find(|x| !x.1.processed)
+                .unwrap();
+            if processed.contains(&obs.0) {
+                // We've processed this point so can move on
+                continue;
+            }
+            processed.insert(obs.0);
+            let neighbors = find_neighbors(&obs.1.observation, observations, self.tolerance());
+            let n = obs.1;
+            n.set_core_distance(self.minimum_points(), &neighbors);
+            result.orderings.push(n.sample());
```

Comment:
does this actually works, how can you ensure that `n` outlives `n.samples`?

Reply:
I figured the lifetime ensured that but I might be wrong

Reply:
if it compiles then it should work :eyes: